Kudu.Services/Diagnostics/ProfileManager.cs

```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
+                    if (profileResponseStatus.StatusCode != HttpStatusCode.OK)
+                    {
+                        return profileResponseStatus;
+                    }
+                }
+                else
                 {
-                    return profileProcessResponse;
+                    string arguments = System.Environment.ExpandEnvironmentVariables(string.Format("start {0} /attach:{1} /loadAgent:{2};DiagnosticsHub.CpuAgent.dll  /scratchLocation:%LOCAL_EXPANDED%\\Temp", profilingSessionId, processId, DIAGNOSTICS_HUB_AGENT_GUID));
+
+                    var profileProcessResponse = await ExecuteProfilingCommandAsync(arguments, tracer);
+
+                    if (profileProcessResponse.StatusCode != HttpStatusCode.OK)
+                    {
+                        return profileProcessResponse;
+                    }
                 }
 
                 // This may fail if we got 2 requests at the same time to start a profiling session
                 // in that case, only 1 will be added and the other one will be stopped.
-                if (!_profilingList.TryAdd(processId, new ProfileInfo(profilingSessionId)))
+                if (!_profilingList.TryAdd(processId, new ProfileInfo(profilingSessionId,iisProfiling)))
                 {
                     tracer.TraceWarning("A profiling session was already running for process {0}, stopping profiling session {1}", processId, profilingSessionId);
-                    await StopProfileInternalAsync(processId, profilingSessionId, true, tracer);
+                    await StopProfileInternalAsync(processId, profilingSessionId, true, tracer, iisProfiling);
                     return new ProfileResultInfo(HttpStatusCode.OK, string.Empty);
                 }
 
                 tracer.Step("started session id: {0} for pid: {1}", profilingSessionId, processId);
 
-                EnsureIdleTimer();
+                EnsureIdleTimer(iisProfiling);
 
                 return new ProfileResultInfo(HttpStatusCode.OK, string.Empty);
             }
         }
 
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
+        {
+            string arguments = System.Environment.ExpandEnvironmentVariables(string.Format("start {0} /scratchLocation:\"%LOCAL_EXPANDED%\\Temp\" /loadAgent:{1};ServiceProfilerAgent.dll /monitor", profilingSessionId, SERVICE_PROFILER_AGENT_GUID));
+            var command = Task.Run(async () =>
+            {
+                await ExecuteProfilingCommandAsync(arguments, tracer);
+            });
+
+            // implementing a retry logic because it can so happen
+            // that the intial profiler command may take some time to start
+            // so max we will wait till 9 seconds before giving up
+            int retryCount = 0;
```

Comment:
There should really be no retry logic here. Tasks can Wait on completion of a task or ContinueWith when a task finishes.